Feature unavailable errors

RonEld opened the pull request:
## Description
1. Add a common error code for alternative implementations, returning `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED`.
2. Remove all feature specific errors for Feature "Not Supported and for the HW acceleration failure.

Dependent on #1993 to be merged first
One of the PRs that supersedes #1716

## Status
**HOLD**

## Requires Backporting
NO  

## Migrations
If there is any API change, what's the incentive and logic for it.

YES 

Removing existing error codes may result n backwards incompatibility. Developers May need to change their returned error codes for their alternative implementations. In addition, If a specific error code is expected, it should be changed as well.
The incentive is to have one common error code for a feature that is not supported by the underlying HW accelerator, which makes tests and applications simpler and readable.

This may result in future ABI break, since removed error codes might be reused in the future,

## Additional comments
Any additional information that could be of interest

The following FEATURE UNAVAILABLE errors were kept, as they actually mean that a specific feature is currently unavailable, because of missing configuration or other reasons, but are not related to the alternative implementation:
~~~
MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE
MBEDTLS_ERR_PKCS12_FEATURE_UNAVAILABLE
MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE
MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE
MBEDTLS_ERR_PEM_FEATURE_UNAVAILABLE
MBEDTLS_ERR_X509_FEATURE_UNAVAILABLE
MBEDTLS_ERR_MD_FEATURE_UNAVAILABLE
MBEDTLS_ERR_PKCS5_FEATURE_UNAVAILABLE
MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE
~~~
In addition, `MBEDTLS_ERR_SSL_HW_ACCEL_FAILED` was kept as it is a specific TLS feature.

Open question: should have `MBEDTLS_ERR_THREADING_FEATURE_UNAVAILABLE` been removed?

## Todos
- [ ] Tests
- [ ] Documentation
- [x] Changelog updated
- [ ] Backported

Patater commented on include/mbedtls/error.h:
```diff
@@ -50,50 +50,42 @@
+ * GCM       2  0x0012-0x0014
+ * BLOWFISH  2  0x0016-0x0018
+ * THREADING 3  0x001C-0x001E
+ * AES       3  0x0020-0x0022   0x0021-0x0021
```
This looks a bit funny. Isn't `0x0021` within the range of `0x0020-0x0022` Why does it need a separate entry?

RonEld replied:
The left is the even error codes, the right are the odd error codes

Patater replied:
Weird. `THREADING` doesn't do that. `BLOWFISH` doesn't do that.

Patater replied:
Nevermind. `BLOWFISH` doesn't do this because we deleted `0x0017`. `THREADING` also doesn't have odd error codes.

Patater commented on include/mbedtls/error.h:
```diff
@@ -50,50 +50,42 @@
+ * BLOWFISH  2  0x0016-0x0018
+ * THREADING 3  0x001C-0x001E
+ * AES       3  0x0020-0x0022   0x0021-0x0021
+ * CAMELLIA  2  0x0024-0x002
```
`0x002` looks wrong.

RonEld replied:
yes, it's a typo

Patater commented on include/mbedtls/gcm.h:
```diff
@@ -146,9 +145,9 @@ int mbedtls_gcm_setkey( mbedtls_gcm_context *ctx,
- * \return          #MBEDTLS_ERR_GCM_HW_ACCEL_FAILED or a cipher-specific
- *                  error code if the encryption or decryption failed.
+ * \return          #MBEDTLS_ERR_GCM_BAD_INPUT if the lengths are not valid or
+ *                  a cipher-specific error code if the encryption
```
A platform error code is not a cipher-specific error code. What makes sense to say here? Should we list `MBEDTLS_ERR_PLATFORM_HW_FAILED` explicitly or say "a cipher- or platform-specific error code"? Same goes for other similar comments for other functions and files.

RonEld replied:
Correct, but the `MBEDTLS_ERR_PLATFORM_HW_FAILED` is a generic error that is returned by the platform. The cipher specific code is because `gcm` is part of the cipher module.

Patater replied:
Where in this function's documentation do we indicate that you might receive a platform error code? Or is that universally implied or expressly written elsewhere?

RonEld replied:
It should be universally implied. It will be updated in the relevant KB article, about implementing alternative modules

Patater replied:
OK

RonEld commented:
@Patater I addressed your comments

RonEld commented:
CI failed in timing test, which is a known issue

RonEld commented:
@Patater Thank you for your review.
Do you have additional comments? Is this approved for design?

RonEld commented:
@Patater @itayzafrir Thank you for reviewing.

I would like to hear your comments about the additional comments in the description.

Patater commented:
> Should MBEDTLS_ERR_THREADING_FEATURE_UNAVAILABLE have been removed?

I could see that making sense, but it's also OK as-is, IMO.

RonEld commented on include/mbedtls/platform.h:
```diff
@@ -43,6 +43,9 @@
 #endif
 
+#define MBEDTLS_ERR_PLATFORM_HW_FAILED           -0x0080 /**< Hardware failed platform operation. */
+#define MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED -0x0082 /**< The requested feature is not supported by the platform */
```
The value probably needs to be changed.
I get the following:
~~~
programs/util/strerror 0x82
Last error was: -0x0082 - UNKNOWN ERROR CODE (0080) : BIGNUM - An error occurred while reading from or writing to a file
~~~
I'll change to 0x78, assuming the parsing will work

RonEld replied:
The problem is that our low level error codes are defined as:
~~~
(0x0002-0x007E, 0x0003-0x007F)
~~~

RonEld commented:
@Patater @itayzafrir I have updated the error code value.
Please review

itayzafrir commented:
Looks good to me. Note there's a typo in e9afdd2, _rage_ -> _range_.

RonEld commented:
@itayzafrir Thanks. But it could also be rage:)

I will fix this once #1993 will be merged, as part of rebasing (I believe there will be conflicts).

RonEld commented:
@Patater Should I rebase to resolve conflicts, or do you want to review first?

Patater commented:
Please rebase and I'll review again from the start. Thanks.

Patater commented:
#1993 has been merged. Removing needs preceding PR label.

RonEld commented:
@Patater I have rebased to resolve conflicts.
I have also fixed the typo in the commit message, mentioned by @itayzafrir iin https://github.com/ARMmbed/mbedtls/pull/2054#issuecomment-432639094

Please review

RonEld commented:
There are two risks here, as I see it:
1. A user uses one of the removed errors, causing compilation errors in the future.
2. Since the removed error codes are not reserved, a future new error might reuse the value, thus causing unexpected behavior when a user expects a value, but different meaning ( could break ABI).

However, a user will probably not use a value number, but an error name, so if there are conflicts, there are probably some compilation errors, and user will use the updated error value.

Patater commented:
Given that these errors are only returned by ALT interfaces, which are non-default configurations of Mbed TLS, this is OK. We will need to update ALT implementations to use the new errors in Mbed OS.

Patater commented:
CI failure is FreeBSD timing test only

RonEld commented:
This PR has been reviewed by two reviewers, so removing the "needs review" label

RonEld commented:
cc @Patater

Patater commented:
Needs a bit of rework for API compatibility's sake

RonEld commented:
As discussed offline, I have returned the removed error codes, for API compatibility sake, and made them deprecated
cc @Patater @itayzafrir

Patater commented on include/mbedtls/aes.h:
```diff
@@ -70,6 +70,17 @@
+#define MBEDTLS_DEPRECATED
+#endif
+
+#define MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE static inline int MBEDTLS_DEPRECATED( void ){ return( MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED ); }
```
- What's the name of this inline function?
- Why do we need an inline function? Add a comment.
- I think I'd rather update our test scripts to work with constant deprecation than to use inline functions to work around the scripts' limitations.

RonEld replied:
>What's the name of this inline function?
Copy paste error. I'll add names to the functions

>Why do we need an inline function? Add a comment.

I couldn't find another way to deprecate MACROs. If I add a warning, we get compilation warnings, and thus compilation errors when warnings are treated as errors. see https://stackoverflow.com/questions/2681259/how-to-deprecate-a-macro-in-gcc

>I think I'd rather update our test scripts to work with constant deprecation than to use inline functions to work around the scripts' limitations.

Me too, but how do you suggest we make the constant deprecations?

RonEld replied:
Note that `all.sh` checks explicitely for `MBEDTLS_DEPRECATED_WARNINGS` and treat warnings as errors:
~~~
msg "build: make, full config + DEPRECATED_WARNING, gcc -O" # ~ 30s
cleanup
cp "$CONFIG_H" "$CONFIG_BAK"
scripts/config.pl full
scripts/config.pl set MBEDTLS_DEPRECATED_WARNING
# Build with -O -Wextra to catch a maximum of issues.
make CC=gcc CFLAGS='-O -Werror -Wall -Wextra' lib programs
make CC=gcc CFLAGS='-O -Werror -Wall -Wextra -Wno-unused-function' tests

msg "build: make, full config + DEPRECATED_REMOVED, clang -O" # ~ 30s
# No cleanup, just tweak the configuration and rebuild
make clean
scripts/config.pl unset MBEDTLS_DEPRECATED_WARNING
scripts/config.pl set MBEDTLS_DEPRECATED_REMOVED
# Build with -O -Wextra to catch a maximum of issues.
make CC=clang CFLAGS='-O -Werror -Wall -Wextra' lib programs
make CC=clang CFLAGS='-O -Werror -Wall -Wextra -Wno-unused-function' tests
~~~

RonEld replied:
@Patater How about we go with our original plan and define the MACROS as the platform errors, but not using the `MBEDTLS_DEPRECATED_WARNING`?

RonEld commented:
As discussed offline, I dropped the deprecation of the constant, and kept the commit for the build error

cc @Patater @itayzafrir

Patater commented:
CI failure is timing-suite on FreeBSD, a known issue.

Patater commented:
Rebased to add back in and deprecate the deprecated errors. Updated ChangeLog wording slightly.

Old branch at https://github.com/Patater/mbedtls/tree/dev/RonEld/feature_unavailable_errors

Patater commented:
Fixed in original commit

sbutcher-arm commented on ChangeLog:
```diff
@@ -2,6 +2,10 @@ mbed TLS ChangeLog (Sorted per branch, date)
 = mbed TLS x.x.x branch released xxxx-xx-xx
 
+API Changes
+   * Add a common error code for a feature that is not supported by the
```
I think we can be specific and say `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED` has been added,

Patater replied:
Yes, specific is better.

sbutcher-arm commented on ChangeLog:
```diff
@@ -24,6 +28,12 @@ Features
 
+New deprecations
+   * All the current module specific errors that mean a feature is not available
+     are deprecated, so the platform error should be used.
```
I think we can specify what we mean here. eg. 
> All the current module specific errors that mean a feature is not available, such as `MBEDTLS_ERR_XXX_FEATURE_UNAVAILABLE`, are deprecated.

sbutcher-arm commented on ChangeLog:
```diff
@@ -24,6 +28,12 @@ Features
+   * All the current module specific errors that mean a feature is not available
+     are deprecated, so the platform error should be used.
+   * All the module specific generic hardware accelaration errors that existed
+     are deprecated, so the platform error should be used.
```
Same comment as above. We can show an example symbol,

sbutcher-arm commented on include/mbedtls/aes.h:
```diff
@@ -60,7 +60,11 @@
 /* Error codes in range 0x0021-0x0025 */
 #define MBEDTLS_ERR_AES_BAD_INPUT_DATA                    -0x0021  /**< Invalid input data. */
+
+/* MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE is deprecated and should not be used. */
```
Not a blocker because there's precedent, but why have the statement on the symbol being deprecated above it and _not in the doxygen comment_?

Patater replied:
Not sure what the best way for our library to document or tell the compiler that a define is deprecated, yet. Copied the approach used previously.

RonEld replied:
If there is a deprecated feature for doxygen, I think it is better. The approach use previously was on a file that is not part of the doxygen headers.

Patater commented:
Yes, please feel free to update the ChangeLog as desired.

sbutcher-arm commented:
retest
